cupy/lib/shape_base.py

```diff
@@ -0,0 +1,67 @@
+from numpy.lib import index_tricks
+
+import cupy
+from cupy import _util
+
+
+def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    """Apply a function to 1-D slices along the given axis.
+
+    Args:
+        func1d (function (M,) -> (Nj...)): This function should accept 1-D
+            arrays. It is applied to 1-D slices of ``arr`` along the specified
+            axis. It must return a 1-D ``cupy.ndarray``.
+        axis (integer): Axis along which ``arr`` is sliced.
+        arr (cupy.ndarray (Ni..., M, Nk...)): Input array.
+        args: Additional arguments for ``func1d``.
+        kwargs: Additional keyword arguments for ``func1d``.
+
+    Returns:
+        cupy.ndarray: The output array. The shape of ``out`` is identical to
+            the shape of ``arr``, except along the ``axis`` dimension. This
+            axis is removed, and replaced with new dimensions equal to the
+            shape of the return value of ``func1d``. So if ``func1d`` returns a
+            scalar ``out`` will have one fewer dimensions than ``arr``.
+
+    .. seealso:: :func:`numpy.apply_over_axes`
+    """
+    ndim = arr.ndim
+    axis = _util._normalize_axis_index(axis, ndim)
+    inarr_view = cupy.moveaxis(arr, axis, -1)
+
+    # compute indices for the iteration axes, and append a trailing ellipsis to
+    # prevent 0d arrays decaying to scalars
+    inds = index_tricks.ndindex(inarr_view.shape[:-1])
+    inds = (ind + (Ellipsis,) for ind in inds)
+
+    # invoke the function on the first item
+    try:
+        ind0 = next(inds)
+    except StopIteration:
+        raise ValueError(
+            'Cannot apply_along_axis when any iteration dimensions are 0'
+        )
+    # cupy.asarray needed in case func1d returns a scalar
+    res = cupy.asarray(func1d(inarr_view[ind0], *args, **kwargs))
```

Comment:
If `func1d` returns a `numpy.ndarray`, it will be converted to cupy.ndarray in this line. However, `ValueError` will be raised in L64 only if the length of the reduction axis is greater than 1.
How about checking if the return value of `func1d` is scalar or cupy.ndarray value?

Reply:
See if the change in 3f6ec11 matches what your were suggesting? Now if `func1d` returns a non-scalar numpy array, a value ValueError (`"non-scalar numpy.ndarray cannot be used for fill"`) will occur when trying to assign the NumPy result to `buff`.